browser/ephemeral_storage/ephemeral_storage_browsertest.cc

```diff
@@ -0,0 +1,302 @@
+  // original value that was written in the non-ephemeral storage area.
+  ValuesFromFrames first_party_values = GetValuesFromFrames(first_party_tab);
+  EXPECT_EQ("b.com - first party", first_party_values.main_frame.local_storage);
+  EXPECT_EQ("b.com - first party", first_party_values.iframe_1.local_storage);
+  EXPECT_EQ("b.com - first party", first_party_values.iframe_2.local_storage);
+
+  EXPECT_EQ("b.com - first party",
+            first_party_values.main_frame.session_storage);
+  EXPECT_EQ("b.com - first party", first_party_values.iframe_1.session_storage);
+  EXPECT_EQ("b.com - first party", first_party_values.iframe_2.session_storage);
+
+  // Even though this page loads b.com iframes as third-party iframes, the TLD
+  // differs, so it should get an entirely different ephemeral storage area.
+  ValuesFromFrames site_c_tab_values = GetValuesFromFrames(site_c_tab);
+  EXPECT_EQ(nullptr, site_c_tab_values.main_frame.local_storage);
+  EXPECT_EQ(nullptr, site_c_tab_values.iframe_1.local_storage);
+  EXPECT_EQ(nullptr, site_c_tab_values.iframe_2.local_storage);
+
+  EXPECT_EQ(nullptr, site_c_tab_values.main_frame.session_storage);
+  EXPECT_EQ(nullptr, site_c_tab_values.iframe_1.session_storage);
+  EXPECT_EQ(nullptr, site_c_tab_values.iframe_2.session_storage);
+}
+
+IN_PROC_BROWSER_TEST_F(EphemeralStorageBrowserTest,
+                       NavigatingClearsEphemeralStorage) {
+  AllowAllCookies();
+
+  ui_test_utils::NavigateToURL(browser(), a_site_ephemeral_storage_url_);
+  auto* web_contents = browser()->tab_strip_model()->GetActiveWebContents();
+
+  SetValuesInFrames(web_contents, "a.com value", "from=a.com");
+
+  ValuesFromFrames values_before = GetValuesFromFrames(web_contents);
+  EXPECT_EQ("a.com value", values_before.main_frame.local_storage);
+  EXPECT_EQ("a.com value", values_before.iframe_1.local_storage);
+  EXPECT_EQ("a.com value", values_before.iframe_2.local_storage);
+
+  EXPECT_EQ("a.com value", values_before.main_frame.session_storage);
+  EXPECT_EQ("a.com value", values_before.iframe_1.session_storage);
+  EXPECT_EQ("a.com value", values_before.iframe_2.session_storage);
+
+  // Navigate away and then navigate back to the original site.
+  ui_test_utils::NavigateToURL(browser(), b_site_ephemeral_storage_url_);
+  ui_test_utils::NavigateToURL(browser(), a_site_ephemeral_storage_url_);
+
+  ValuesFromFrames values_after = GetValuesFromFrames(web_contents);
+  EXPECT_EQ("a.com value", values_after.main_frame.local_storage);
+  EXPECT_EQ(nullptr, values_after.iframe_1.local_storage);
+  EXPECT_EQ(nullptr, values_after.iframe_2.local_storage);
+
+  EXPECT_EQ("a.com value", values_after.main_frame.session_storage);
+  EXPECT_EQ(nullptr, values_after.iframe_1.session_storage);
+  EXPECT_EQ(nullptr, values_after.iframe_2.session_storage);
+}
+
+IN_PROC_BROWSER_TEST_F(EphemeralStorageBrowserTest,
+                       ClosingTabClearsEphemeralStorage) {
```

Comment:
maybe also check that reload doesn't drop the storage?

Reply:
and also check that storage is different in another profile (should be relatively easy to spinoff an incognito window and do same checks)

Reply:
(both tests can be a follow-up of course)

Reply:
Okay. This all makes sense and I'm happy to add these tests in a quick followup, as you suggest.